File: lab10/i.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> g(n+1);
    vector<int> indeg(n+1, 0);

    for (int k = 0; k < m; k++) {
        int i, j;
        cin >> i >> j;
        g[i].push_back(j);
        indeg[j]++;
    }

    queue<int> q;

    for (int i = 1; i <= n; i++) {
        if (indeg[i] == 0)
            q.push(i);
    }

    vector<int> order;

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        order.push_back(u);

        for (int v : g[u]) {
            indeg[v]--;
            if (indeg[v] == 0)
                q.push(v);
        }
    }

    if ((int)order.size() == n) {
        cout << "Possible\n";
        for (int x : order)
            cout << x << " ";
        cout << "\n";
    } else {
        cout << "Impossible\n";
    }

    return 0;
}

File: lab10/i.py
import sys 
input = sys.stdin.readline 
from collections import deque 

n , m = map(int , input().split())

g = [[] for _ in range(n+1)]
indeg = [0] * (n+1)

for _ in range(m):
    i , j = map(int , input().split())
    g[i].append(j)
    indeg[j] += 1 

q = deque() 
for i in range(1 , n+1):
    if indeg[i] == 0:
        q.append(i)

order = []

while q:
    u = q.popleft() 
    order.append(u)

    for v in g[u]:
        indeg[v] -= 1 
        if indeg[v] == 0:
            q.append(v) 

if len(order) == n:
    print("Possible")
    print(*order)
else:
    print("Impossible")